compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -285,6 +286,18 @@ public boolean isSubTypeOfList(BType type) {
         return ((BUnionType) type).getMemberTypes().stream().allMatch(this::isSubTypeOfList);
     }
 
+    public BType resolvePatternTypeFromMatchExpr(BType matchExprType, BType listMatchPatternType,
+                                                 BLangListMatchPattern listMatchPattern) {
+
+        if (isAssignable(listMatchPatternType, matchExprType)) {
+            return listMatchPatternType;
+        }
+        if (isAssignable(matchExprType, listMatchPatternType)) {
+            return matchExprType;
+        }
```

Comment:
I'm not sure if just these two checks will be enough. Shouldn't we consider intersections?

Shouldn't the following be allowed?

```ballerina
public function main() {
    [any, any] m = [1, 2];
    int|error a;
    int|error b;

    match m {
        [a, b] => {
            
        }
    }
}
```

Currently fails with
```cmd
$ ballerina run zxa.bal Compiling source
        zxa.bal
error: .::zxa.bal:7:9: pattern will not be matched
```

Reply:
Can't we use org.wso2.ballerinalang.compiler.semantics.analyzer.Types#getTypeIntersection if that is the case?

Reply:
Added `getTypeIntersection` function. Above example is not supported according to spec, because the var ref should be declared a module-level const. Anyway, the error message is incorrect and it is fixed.

Reply:
Ah, yes. Tested with the current changes and the errors also seem to be correct.

```cmd
error: .::xyz.bal:7:10: variable 'a' should be declared as constant
error: .::xyz.bal:7:13: variable 'b' should be declared as constant
```